mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h

```diff
@@ -0,0 +1,421 @@
+/// Each column represents the memory access pattern of the corresponding loop
+/// level (the outermost loop maps to the leftmost column in the matrix).
+/// Each row represents the memory access pattern of the corresponding array
+/// dimension (the leftmost dimension maps to first row in the matrix).
+/// Note: The array access function is expected to be a linear combination of
+/// the enclosing loop induction variables.
+class MemoryAccessMatrix {
+  friend raw_ostream &operator<<(raw_ostream &, const MemoryAccessMatrix &);
+
+public:
+  MemoryAccessMatrix() = delete;
+  MemoryAccessMatrix(MemoryAccessMatrix &&) = default;
+  MemoryAccessMatrix(const MemoryAccessMatrix &) = default;
+
+  /// Construct a matrix with the specified number of rows and columns.
+  MemoryAccessMatrix(unsigned nRows, unsigned nColumns);
+
+  /// Construct a matrix from a initializer list.
+  MemoryAccessMatrix(
+      std::initializer_list<std::initializer_list<Value>> initList);
+
+  /// Access the element at the specified \p row and \p column.
+  Value &at(unsigned row, unsigned column) {
+    assert(row < nRows && "Row outside of range");
+    assert(column < nColumns && "Column outside of range");
+    return data[row * nColumns + column];
+  }
+
+  Value at(unsigned row, unsigned column) const {
+    assert(row < nRows && "Row outside of range");
+    assert(column < nColumns && "Column outside of range");
+    return data[row * nColumns + column];
+  }
+
+  Value &operator()(unsigned row, unsigned column) { return at(row, column); }
+  Value operator()(unsigned row, unsigned column) const {
+    return at(row, column);
+  }
+
+  /// Swap \p column with \p otherColumn.
+  void swapColumns(unsigned column, unsigned otherColumn);
+
+  /// Swap \p row with \p otherRow.
+  void swapRows(unsigned row, unsigned otherRow);
+
+  unsigned getNumRows() const { return nRows; }
+
+  unsigned getNumColumns() const { return nColumns; }
+
+  /// Get a copy of the specified \p row.
+  SmallVector<Value> getRow(unsigned row) const;
+
+  /// Set the specified \p row to \p elems.
+  void setRow(unsigned row, ArrayRef<Value> elems);
+
+  /// Fill \p row with the given value \p val.
+  void fillRow(unsigned row, Value val);
+
+  /// Add an extra row at the bottom of the matrix.
+  unsigned appendRow();
+
+  /// Add an extra row at the bottom of the matrix and copy the given elements
+  /// \p elems into the new row.
+  unsigned appendRow(ArrayRef<Value> elems);
+
+  /// Get a copy of the specified \p column.
+  SmallVector<Value> getColumn(unsigned column) const;
+
+  /// Set the specified \p column to \p elems.
+  void setColumn(unsigned col, ArrayRef<Value> elems);
+
+  /// Fill \p col with the given value \p val.
+  void fillColumn(unsigned col, Value val);
+
+  /// Fill the matrix with the given value \p val.
+  void fill(Value val);
+
+  /// Construct a new matrix containing the specified \p rows.
+  MemoryAccessMatrix getRows(std::set<unsigned> rows) const;
```

Comment:
I would expect `getRows({2, 0})` to return the third row first, then the first row, using `std::set` would be wrong in that case, how about using `SmallSetVector`?